### PR TITLE
Improvements

### PR DESCRIPTION
**Pull request by cxhernandez**
I've attempted to make this a proper python package and optimize your original codebase using Cython. To get the persistence from a dataset, the workflow is also now more `sklearn`-like:

```python
import numpy as np
from tda import PersistentHomology

data = np.array([[1,4],[1,1],[6,1],[6,4]])

ph = PersistentHomology(epsilon=7.)
persist = ph.fit_transform(data)
```

which takes just over a millisecond to compute.

For the figure 8 example (w/ jitter added), the code now clocks in around ~120 milliseconds.

```
In [8]: ph = PersistentHomology(epsilon=0.2)
In [9]: %timeit ph.fit_transform(data)
10 loops, best of 3: 126 ms per loop
```

I'll try to add docstrings and such later.

Thanks for all your original hard work!

### CODE REVIEW

File: tda/persistent_homology.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+import networkx as nx
+from scipy.spatial.distance import squareform, pdist
+
+from .snf import low, smith_normal_form
+
+__all__ = ['PersistentHomology']
+
+
+def buildGraph(data, epsilon=1., metric='euclidean', p=2):
+    D = squareform(pdist(data, metric=metric, p=p))
```
Comment (cxhernandez) [inline tda/persistent_homology.py]: This could also be improved by using sparse matrices and `cKDTree`.

**Comment by outlace** [comment]
This is very neat! Thanks for contributing. I will merge it. I will add a /legacy/ folder with the original code because this was originally meant to be associated with my blog's tutorial, so I want blog readers to still be able to follow along with the original code.

I'm also planning to add the Mapper algorithm to this project. I know the original Mapper is a Python project but it's not as usable as I would like.

**Comment by cxhernandez** [comment]
I found an alternative implementation of the mapper algorithm [here](https://github.com/szairis/sakmapper) in pure Python, in case you're interested.

**Comment by outlace** [comment]
That project hasn't been updated in 4 months despite saying it's in "active development."
If we add Mapper to this project, we can have a complete open-source TDA library with both major tools, PH and Mapper. We could pull code form that other project so we don't need to build it from scratch, however. 

If you're interested in contributing, then we can make this into a useful performant TDA library for Python.

**Pull request merged**